tsweb/jsonhandler.go

```diff
@@ -0,0 +1,280 @@
+// Copyright (c) 2020 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package tsweb
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"mime"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"tailscale.com/types/logger"
+)
+
+// JSONHandler is like net/http.Handler, but the handler has JSON and Form
+// data parsed for it out of http.Request and returns JSON or an error
+// that is converted to JSON instead of writing to its ResponseWriter.
+type JSONHandler interface {
+	// ServeHTTPJSON is like http.Handler.ServeHTTP,
+	// except optimized for JSON REST APIs.
+	//
+	// If the request is a POST, then the request data is decoded
+	// and passed as a map to ServeHTTPJSON. The data can be either
+	// a url-encoded form or a JSON blob. The JSON blob is required
+	// to be a property set (that is, start with '{').
+	//
+	// The caller will write a response to the ResponseWriter.
+	// The response is always a JSON blob, either:
+	//
+	//	{"success": true, "data": <encoded obj>}
+	//
+	// or
+	//
+	//	{"success": false, "error": "<error message>"}
+	//
+	// If ServeHTTPJSON returns no error, the returned obj is
+	// encoded as a JSON property list and sent to the client.
+	//
+	// If ServeHTTPJSON returns an error, it caller should handle
+	// an error by serving an HTTP 500 response to the user. The
+	// error details must not be sent to the client, as they may
+	// contain sensitive information. If the error is an
+	// HTTPError, though, callers should use the HTTP response
+	// code and message as the response to the client.
+	ServeHTTPJSON(http.ResponseWriter, *http.Request, map[string]interface{}) (obj interface{}, err error)
+}
+
+// StdJSONHandler converts a JSONHandler into a standard http.Handler.
+// Handled requests are logged using logf, as are any errors. Errors
+// are handled as specified by the Handler interface.
+func StdJSONHandler(jh JSONHandler, logf logger.Logf) http.Handler {
+	return jsonHandler{jh: jh, logf: logf, timeNow: time.Now, log200s: true}
+}
+
+// JSONHandlerFunc is an adapter to allow the use of ordinary
+// functions as JSONHandlers. If f is a function with the
+// appropriate signature, JSONHandlerFunc(f) is a JSONHandler that
+// calls f.
+//
+// See the documentation on JSONHandler.ServeHTTPJSON for semantics
+// of the parameters.
+type JSONHandlerFunc func(http.ResponseWriter, *http.Request, map[string]interface{}) (interface{}, error)
```

Comment:
I removed this once (in 64334143a1e1c83fc5392422bf28f0aa701e6211) and Dave added it back (in eccae0cd) but I really don't want either of them. It's been on my list to remove it again.

The composable base unit of net/http is the http.Handler. HandlerFunc just gets you to that. Once you have new types like ReturnHandler and JSONHandler they stop composing with each other. Adding these adapters are helping you get to those types, rather than help you get to an http.Handler. (this is all related to the mistakes we made & learned from http.Flusher, http.Pusher, etc)

That's why I'd rather not even have a JSONHandler type and instead have something that did let you write the func you want to write, but turn it into an http.Handler.

Reply:
So, writing more code in and around control, these "http.Handler with errors" types are cumbersome to use _unless_ we do actually make them our standard type instead of http.Handler. And, making them the default over http.Handler breaks integration with things like Gorilla middleware, as Brad says.

I'm fairly convinced at this point that http.Handler should have returned an error. Given that it doesn't, I'm no longer convinced of the wisdom of us trying to retrofit it, vs. accomplishing similar goals differently (e.g. have a logging middleware that records the appropriate information and generates request logs, or some kind of awful panic-based error return that doesn't require altering the function signature). I'd love to talk this out with y'all.

Reply:
I suggested earlier today that @lzjluzijie look at implementing the `NewJSONHandler` syntax Brad proposed above if he's looking for a mini-project, as I don't expect to get to it for a couple weeks at my current rate. I think it's going to be generally pretty useful for JSON handlers and I'm not sure the the gorilla middleware has convinced me otherwise. If you think so, we should chat about it before Zijie puts any time into this.